lib/ddtrace/context_flush.rb

```diff
@@ -1,132 +1,69 @@
-require 'set'
-
-require 'ddtrace/context'
-
 module Datadog
-  # \ContextFlush is used to cap context size and avoid it using too much memory.
-  # It performs memory flushes when required.
-  class ContextFlush
-    # by default, soft and hard limits are the same
-    DEFAULT_MAX_SPANS_BEFORE_PARTIAL_FLUSH = Datadog::Context::DEFAULT_MAX_LENGTH
-    # by default, never do a partial flush
-    DEFAULT_MIN_SPANS_BEFORE_PARTIAL_FLUSH = Datadog::Context::DEFAULT_MAX_LENGTH
-    # timeout should be lower than the trace agent window
-    DEFAULT_PARTIAL_FLUSH_TIMEOUT = 10
-
-    private_constant :DEFAULT_MAX_SPANS_BEFORE_PARTIAL_FLUSH
-    private_constant :DEFAULT_MIN_SPANS_BEFORE_PARTIAL_FLUSH
-    private_constant :DEFAULT_PARTIAL_FLUSH_TIMEOUT
-
-    def initialize(options = {})
-      # max_spans_before_partial_flush is the amount of spans collected before
-      # the context starts to partially flush parts of traces. With a setting of 10k,
-      # the memory overhead is about 10Mb per thread/context (depends on spans metadata,
-      # this is just an order of magnitude).
-      @max_spans_before_partial_flush = options.fetch(:max_spans_before_partial_flush,
-                                                      DEFAULT_MAX_SPANS_BEFORE_PARTIAL_FLUSH)
-      # min_spans_before_partial_flush is the minimum number of spans required
-      # for a partial flush to happen on a timeout. This is to prevent partial flush
-      # of traces which last a very long time but yet have few spans.
-      @min_spans_before_partial_flush = options.fetch(:min_spans_before_partial_flush,
-                                                      DEFAULT_MIN_SPANS_BEFORE_PARTIAL_FLUSH)
-      # partial_flush_timeout is the limit (in seconds) above which the context
-      # considers flushing parts of the trace. Partial flushes should not be done too
-      # late else the agent rejects them with a "too far in the past" error.
-      @partial_flush_timeout = options.fetch(:partial_flush_timeout,
-                                             DEFAULT_PARTIAL_FLUSH_TIMEOUT)
-      @partial_traces = []
+  module ContextFlush
+    # Consumes only completed traces (where all spans have finished)
+    class Finished
+      # Consumes and returns completed traces (where all spans have finished)
+      # from the provided +context+, if any.
+      #
+      # Any traces consumed are removed from +context+ as a side effect.
+      #
+      # @return [Array<Span>] trace to be flushed, or +nil+ if the trace is not finished
+      def consume(context)
```

Comment:
I see... consume is meant to imply there's side effects (as you can't "consume" something multiple times.") The description helps clarify this.

I might suggest considering adding `!` in a way that's Ruby convention, but this isn't a major point of feedback since this isn't a public facing API.

Reply:
Added ! to consume method.